interfaces/builtin/cups.go

```diff
@@ -48,15 +58,108 @@ const cupsConnectedPlugAppArmor = `
 
 #include <abstractions/cups-client>
 /{,var/}run/cups/printcap r,
+
+# allow talking to the snap cups socket
+/var/cups/cups.sock rw,
 `
 
+type cupsInterface struct {
+	commonInterface
+}
+
+func (iface *cupsInterface) AppArmorConnectedSlot(spec *apparmor.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	return nil
+}
+
+func (iface *cupsInterface) BeforePrepareSlot(slot *snap.SlotInfo) error {
+	// verify that the snap has a cups-socket interface attribute, which is
+	// needed to identify where to find the cups socket is located in the snap
+	// providing the cups socket
+
+	var cupsdSocketSource string
+	if err := slot.Attr("cups-socket", &cupsdSocketSource); err != nil {
+		return err
+	}
+
+	// TODO: we probably want to allow the empty slot for old cups snaps which
+	//       won't have the slot and have a sensible default for those, or
+	//       possibly just don't emit the mount if the attribute is not present
+
+	if cupsdSocketSource == "" {
+		return fmt.Errorf("cups slot must specify the location of the cups socket")
+	}
+
+	// TODO: should we do some checks on the cups-socket file location that it
+	//       is a clean filepath, etc.
+
+	return nil
+}
+
+func (iface *cupsInterface) AppArmorConnectedPlug(spec *apparmor.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	// add the base snippet
+	spec.AddSnippet(cupsConnectedPlugAppArmor)
+
+	// setup the snap-update-ns rules for bind mounting
+	emit := spec.AddUpdateNSf
+
+	var cupsdSocketSource string
+	err := slot.Attr("cups-socket", &cupsdSocketSource)
+	if err != nil {
+		return fmt.Errorf("broken interface: %v", err)
+	}
+
+	cupsdSocketSource = resolveSpecialVariable(cupsdSocketSource, slot.Snap())
+
+	emit("  # Mount cupsd socket from cups snap to client snap\n")
+	emit("  mount options=(bind rw) %s -> /var/cups/cups.sock,\n", cupsdSocketSource)
+	emit("  umount /var/cups/cups.sock,\n")
+
+	apparmor.GenWritableProfile(emit, cupsdSocketSource, 1)
+	apparmor.GenWritableProfile(emit, "/var/cups/cups.sock", 1)
+
+	// TODO: figure out why this bit is necessary - without this, apparmor
+	// denies trying to connect to the unix socket at /var/run/cups.sock with:
+	//
+	// apparmor="DENIED" operation="connect"
+	// profile="snap.test-snapd-cups-consumer.bin"
+	// name="/var/snap/test-snapd-cups-provider/common/cups.sock"
+	// pid=3195747 comm="nc" requested_mask="wr" denied_mask="wr" fsuid=0 ouid=0
```

Comment:
I've seen this same type of error when hard linking a socket file and trying to open the socket from the new path. I haven't checked the source code, but I suspect the AppArmor check is being performed against the path that the server bound their socket to rather than the path the client is trying to connect to.

It's quite possible that things will behave differently with the upstreamed version of the unix socket support.

Reply:
Yes this is actually a known problem, the content interface ran into this same problem and fixed it with the same rule, so I just copied the comment from the content interface here, so it's a known thing we are at least working around consistently across snapd.